File: js/test_money.js
const assert = require('assert');
const Money = require('./money');
const Portfolio = require('./portfolio');


class MoneyTest {
  testAdditionOfDollarsAndWons() {
    let one = new Money(1, "USD")
    let eleven = new Money(1100, "KRW")
    let portfolio = new Portfolio();
    portfolio.add(one, eleven);
    let expected = new Money(2200, "KRW")
    assert.deepStrictEqual(portfolio.evaluate("KRW"), expected);
  }

  testAdditionOfDollarsAndEuros() {
    let five = new Money(5, "USD")
    let ten = new Money(10, "EUR")
    let portfolio = new Portfolio();
    portfolio.add(five, ten)
    let expected = new Money(17, "USD")
    assert.deepStrictEqual(portfolio.evaluate("USD"), expected)
  }

  testMultiplication() {
    let money = new Money(10, "EUR")
    let expected = new Money(20, "EUR")
    assert.deepStrictEqual(money.times(2), expected);
  }

  testDivision() {
    let money = new Money(4002, "KRW")
    let expected = new Money(1000.5, "KRW")
    assert.deepStrictEqual(money.divide(4), expected)
  }

  testAddition() {
    let five = new Money(5, "USD")
    let ten = new Money(10, "USD")
    let fifteen = new Money(15, "USD")
    let portfolio = new Portfolio();
    portfolio.add(five, ten);
    assert.deepStrictEqual(portfolio.evaluate("USD"), fifteen)
  }

  get() {
    let proto = MoneyTest.prototype;
    let props = Object.getOwnPropertyNames(proto);
    let methods = props.filter(p => {
      return typeof proto[p] === "function" && p.startsWith("test");
    });
    return methods;
  }

  run() {
    let testMethods = this.get();
    testMethods.forEach(m => {
      console.log("Running: %s()", m)
      let method = Reflect.get(this, m);
      try {
        Reflect.apply(method, this, []);
      } catch (e) {
        if (e instanceof assert.AssertionError) {
          console.log(e);
        } else {
          throw e;
        }
      }
    })
  }
}

new MoneyTest().run();

File: py/test_money.py
import unittest

from money import Money
from portfolio import Portfolio
from bank import Bank


class TestMoney(unittest.TestCase):
    def setUp(self):
        self.bank = Bank()
        self.bank.add_exchange_rate("EUR", "USD", 1.2)
        self.bank.add_exchange_rate("USD", "KRW", 1100)

    def test_conversion_with_missing_exchange_rate(self):
        bank = Bank()
        ten_euros = Money(10, "EUR")
        with self.assertRaisesRegex(Exception, "EUR->Kalganid"):
            bank.convert(ten_euros, "Kalganid")

    def test_conversion(self):
        bank = Bank()
        bank.add_exchange_rate("EUR", "USD", 1.2)
        ten_euros = Money(10, "EUR")
        self.assertEqual(bank.convert(ten_euros, "USD"), Money(12, "USD"))

    def test_addition_with_multiple_missing_exchange_rates(self):
        one_dollar = Money(1, "USD")
        one_euro = Money(1, "EUR")
        one_won = Money(1, "KRW")
        portfolio = Portfolio()
        portfolio.add(one_dollar, one_euro, one_won)

        with self.assertRaisesRegex(
            Exception,
            r"Missing exchange rate\(s\):\[USD\-\>Kalganid,EUR\-\>Kalganid,KRW\-\>Kalganid\]",
        ):
            portfolio.evaluate(Bank(), "Kalganid")

    def test_addition_of_dollars_and_wons(self):
        one = Money(1, "USD")
        eleven = Money(1100, "KRW")
        portfolio = Portfolio()
        portfolio.add(one, eleven)
        expected = Money(2200, "KRW")
        self.assertEqual(expected, portfolio.evaluate(self.bank, "KRW"))

    def test_addition_of_dollars_and_euros(self):
        five = Money(5, "USD")
        ten = Money(10, "EUR")
        portfolio = Portfolio()
        portfolio.add(five, ten)
        expected = Money(17, "USD")
        result = portfolio.evaluate(self.bank, "USD")
        self.assertEqual(expected, result)

    def test_addition(self):
        five = Money(amount=5, currency="USD")
        ten = Money(amount=10, currency="USD")
        fifteen = Money(amount=15, currency="USD")
        portfolio = Portfolio()
        portfolio.add(five, ten)
        self.assertEqual(fifteen, portfolio.evaluate(self.bank, "USD"))

    def test_multiplication_in_euros(self):
        money = Money(amount=10, currency="EUR")
        expected = Money(amount=20, currency="EUR")
        self.assertEqual(money.times(2), expected)

    def test_division(self):
        money = Money(amount=4002, currency="KRW")
        expected = Money(amount=1000.5, currency="KRW")
        self.assertEqual(money.divide(4), expected)


if __name__ == "__main__":
    unittest.main()
